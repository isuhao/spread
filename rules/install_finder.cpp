#include "install_finder.hpp"
#include <boost/filesystem.hpp>
#include <set>

using namespace Spread;

InstallFinder::InstallFinder(const RuleFinder &_rules, Cache::CacheIndex &_cache)
  : rules(_rules), cache(_cache) {}

void InstallFinder::handleDeps(const DepList &deps, ActionMap &output)
{
  for(int i=0; i<deps.size(); i++)
    {
      const std::string &dest = deps[i].first;
      const Hash &hash = deps[i].second;

      // Check if this hash has already been resolved.
      ActionMap::iterator it = output.find(hash);
      if(it != output.end())
        {
          // Add ourself to it's destination list, if we have one
          it->second.addDest(dest);

          continue;
        }

      // Hash not previously resolved. Check if the cache knows
      // anything about the hash.
      int stat = Cache::CI_ElseWhere;
      if(dest != "")
        stat = cache.getStatus(dest, hash);

      if(stat == Cache::CI_Match)
        {
          /* Add copy action with empty destination. This allows other
             tasks to look up the same action later and add multiple
             destinations.

             If no other actions need this file, then the action is
             ignored, which is OK since the file is already installed.
          */
          output[hash] = Action(dest);
          continue;
        }

      if(stat == Cache::CI_ElseWhere)
        {
          // File either exists elsewhere, or dest == "". Get cache
          // location, if any.
          const std::string &existing = cache.findHash(hash);

          if(existing != "")
            {
              // Yup, it exists. Add a copy operation.
              output[hash] = Action(existing, dest);
              continue;
            }
        }

      /* If we are here, no existing file met our dependency. Look
         up the ruleset to see how else we can fetch the file.
      */
      const Rule *r = rules.findRule(hash);

      if(r)
        {
          bool found = false;

          // Set up the output action
          Action a(r);

          // Loop through and add the action for all the hashes it
          // produces, not just the one we are looking for.
          for(int k=0; k<r->outputs.size(); k++)
            {
              const Hash &h = r->outputs[k];
              output[h] = a;
              if(h == hash)
                {
                  // Add our output destination to the right output
                  output[h].addDest(dest);
                  found = true;
                }
            }
          assert(found);

          // Then loop through this rule's dependencies, and expand
          // those as well
          DepList subdeps;
          for(int k=0; k<r->deps.size(); k++)
            {
              const Hash &h = r->deps[k];
              subdeps.push_back(DepPair("",h));
              handleDeps(subdeps, output);
            }

          /* TODO: Handle recursion. Recursion happens when a hash
             implicitly depends upon itself. We would also have to
             figure out what to do in the case of recursion. At the
             moment though, we just assume our ruleset is acyclic.
          */
          continue;
        }

      // There was no way to resolve this hash. An empty copy action
      // represents an unhandled dependency.
      output[hash] = Action("", dest);
    }
}
