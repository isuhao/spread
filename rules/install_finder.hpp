#ifndef __SPREAD_INSTALL_FINDER_HPP_
#define __SPREAD_INSTALL_FINDER_HPP_

#include "cache/index.hpp"
#include "hash/hash.hpp"
#include "rulefinder.hpp"
#include "actions.hpp"
#include <vector>

namespace Spread
{
  struct InstallFinder
  {
    InstallFinder(const RuleFinder &rules, Cache::CacheIndex &cache);

    /* Add a dependency. A dependency is a request for a specific
       object 'hash' to be installed in location 'file'.

       Several dependencies may overwrite the same location. It's
       always the one added last that takes effect. This can be used
       intentionally to install eg. a game, and then a patch on top of
       it.
     */
    void addDep(const std::string &file, const Hash &hash)
    { deps.push_back(DepPair(file,hash)); }

    /* Perform the rule resolution. Takes the list of input
       dependencies in 'deps' and converts it to a list of actions in
       'output'.
     */
    void perform(ActionMap &output) { handleDeps(deps, output); }

    typedef std::pair<std::string,Hash> DepPair;
    typedef std::vector<DepPair> DepList;

    /* This is the struct's input data. You can set it up through the
       addDep functions above, or manually if you prefer. Then run
       perform() to produce the output.

       Running perform() does not change the depslist, so the struct
       is reusable.
     */
    DepList deps;

  private:
    const RuleFinder &rules;
    Cache::CacheIndex &cache;

    void handleDeps(const DepList &deps, ActionMap &output);
  };
}
#endif
