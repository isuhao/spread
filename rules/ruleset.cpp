#include "ruleset.hpp"
#include <vector>
#include <map>
#include <ctime>
#include "misc/random.hpp"
#include <boost/thread/recursive_mutex.hpp>

using namespace Spread;

typedef boost::shared_ptr<URLRule> URLPtr;
typedef std::vector<URLPtr> UVec;
typedef std::map<Hash, UVec> UMap;

#define LOCK boost::lock_guard<boost::recursive_mutex> lock(ptr->mutex)

struct RuleSet::_RuleSetInternal
{
  UMap urls;

  Misc::Random rnd;
  boost::recursive_mutex mutex;

  // Returns an URL rule or NULL
  const Rule* findURL(const Hash &hash)
  {
    UMap::iterator it = urls.find(hash);
    if(it == urls.end())
      return NULL;

    UVec &vec = it->second;

    if(!vec.size())
      return NULL;

    // First, build a list of rules with the highest priority
    int prio = vec[0]->priority;
    std::vector<int> indices;
    double sum = 0;

    for(int i=0; i<vec.size(); i++)
      {
        const URLRule &r = *vec[i].get();

        // Skip disabled rules
        if(r.isBroken) continue;

        // Drop all previous results if this has a higher priority
        if(r.priority > prio)
          {
            indices.clear();
            prio = r.priority;
            sum = 0;
          }

        // Skip everything below the highest priority level
        else if(r.priority < prio)
          continue;

        // Add this to the weight list
        if(r.weight > 0) sum += r.weight;
        indices.push_back(i);
      }

    if(indices.size() == 0)
      return NULL;

    /* Pick a card, any card!

       Generates a randum number between 0 and 'sum'.
     */
    double pick = rnd.dgen()*sum;

    // Pick the right one
    double psum = 0;
    for(int i=0; i<indices.size(); i++)
      {
        URLPtr ptr = vec[indices[i]];
        float w = ptr->weight;
        if(w <= 0) continue;
        psum += w;

        if(pick < psum)
          return ptr.get();
      }

    /* In case our impeccable math skillz failed, just go for the
       first one. This is also simpler than trying to account for
       pathological corner cases such as NaN weights etc.
    */
    return vec[indices[0]].get();
  }
};

RuleSet::RuleSet()
{
  ptr.reset(new _RuleSetInternal);
}

const Rule *RuleSet::findRule(const Hash &hash) const
{
  LOCK;

  // Search URLs
  const Rule *rule = ptr->findURL(hash);
  if(rule) return rule;

  return NULL;
}

void RuleSet::addURL(const Hash &hash, const std::string &url,
                     int priority, float weight,
                     std::string ruleString)
{
  if(ruleString == "")
    ruleString = "URL " + hash.toString() + " " + url;

  LOCK;
  ptr->urls[hash].push_back
    (URLPtr(new URLRule(hash, ruleString, url, priority, weight)));
}

void RuleSet::reportBrokenURL(const Hash &hash, const std::string &url)
{
  LOCK;

  // Find all matching rules and disable them
  UMap::iterator it = ptr->urls.find(hash);
  if(it != ptr->urls.end())
    {
      UVec &vec = it->second;

      for(int i=0; i<vec.size(); i++)
        {
          URLRule *r = vec[i].get();
          if(r->url == url)
            r->isBroken = true;
        }
    }

  // Invoke the callback to notify external systems about the broken
  // URL.
  if(callback) callback(hash, url);
}

const URLRule *RuleSet::getURL(const Rule *r)
{
  assert(r->type == RST_URL);
  const URLRule *ur = dynamic_cast<const URLRule*>(r);
  assert(ur != NULL);
  return ur;
}
