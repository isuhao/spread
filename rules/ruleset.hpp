#ifndef __SPREAD_RULESET_HPP_
#define __SPREAD_RULESET_HPP_

#include "rulefinder.hpp"
#include <boost/shared_ptr.hpp>
#include <boost/function.hpp>

/* RuleSet is an implementation of the abstract RuleFinder interface.

   The RuleSet holds the global part of the rule set in memory. It
   holds URL and archive rules, but does not index hashes within the
   archives themselves.
 */

namespace Spread
{
  // These are used for the Rule::type field on all produced rules
  enum RuleSetTypes
    {
      RST_None          = 0,    // No/unknown rule type
      RST_URL           = 1,    // URL rule
      RST_Archive       = 2     // Archive / unpack rule
    };

  /* Specific Rule sublclasses for the rule types known to RuleSet.
   */
  struct URLRule : Rule
  {
    std::string url;
    bool isBroken;

    /* URL priorities are used to force some urls to always be chosen
       before others. If there exists more than one URL source for a
       given hash, the one with the highest priority is always
       returned. The only way a lower priority URL may be returned is
       if all the ones above it have failed (reported as broken.)

       This allows setting up fallback URLs to be used in case of
       temporary server failures, for example.

       If multiple sources share the highest priority, then weighting
       is used.
     */
    int priority;

    /* Weighting is used to pick from multiple sources that have the
       same priority. All the weights for the given priority level are
       summed, and each source is given a probability of weight/sum.
       Then a source is picked at random using those probabilities.
     */
    float weight;

    URLRule(const Hash &hash, const std::string &rulestr,
            const std::string &_url, int prio = 1, float wght = 1.0)
      : Rule(RST_URL, rulestr), url(_url), isBroken(false),
        priority(prio), weight(wght) { addOut(hash); }
  };

  struct RuleSet : RuleFinder
  {
    RuleSet();

    /* Inherited from RuleFinder.

       NOTE: Returned pointers are only valid for the lifetime of the
       RuleSet instance. Objects are deleted when RuleSet destructs.
     */
    const Rule *findRule(const Hash &hash) const;
    const Directory *findDir(const Hash &hash) const { return NULL; }

    /* Add an URL rule.
    */
    void addURL(const Hash &hash, const std::string &url,
                int priority = 1, float weight = 1.0,
                std::string ruleString = "");

    /* Decode and add rule string.

       Known formats:

       "URL <hash> <url> [priority] [weight]"
         : URL rule. See URLRule for description of the priority and
           weight fields.

       "ARC <arc-hash> <dir-hash>"
         : Archive rule. Archive in object <arc-hash> unpacks into the
           objects listed in <dir-hash>.
     */
    void addRule(const std::string &ruleString) {} // Not implemented yet

    /* Report broken URL rule. This will remove the rule from all
       future rule searches.

       This will also invoke a callback (if set through
       setURLCallback) which you can use to report the issue back to
       the server, or do other actions to correct the problem.
     */
    void reportBrokenURL(const Hash &hash, const std::string &url);

    // Get URLRule pointer from a Rule pointer
    static const URLRule *getURL(const Rule *r);

    // Set callback to handle broken URLs
    typedef boost::function< void(const Hash &hash, const std::string &url) > CBFunc;
    void setURLCallback(CBFunc cb) { callback = cb; }

  private:
    struct _RuleSetInternal;
    boost::shared_ptr<_RuleSetInternal> ptr;
    CBFunc callback;
  };

}

#endif
