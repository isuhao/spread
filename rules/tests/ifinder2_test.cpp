#include <iostream>
#include "install_finder.hpp"
#include "ruleset.hpp"

using namespace std;
using namespace Spread;

Hash hello("hello",5);
Hash dolly("dolly",5);

void print(InstallFinder &inf)
{
  ActionMap output;
  bool isOk = inf.perform(output);

  cout << "\nHandled " << inf.deps.size() << " input dependencies into "
       << output.size() << " hashes (OK=" << isOk << "):\n";
  ActionMap::iterator it;
  for(it = output.begin(); it != output.end(); it++)
    {
      Action &a = it->second;

      cout << it->first << " : ";

      if(a.isNone())
        {
          cout << "UNHANDLED\n";
        }
      else if(a.isRule())
        {
          const Rule &r = *a.rule;

          cout << r.ruleString << endl;

          for(int i=0; i<r.deps.size(); i++)
            cout << "  <= " << r.deps[i] << endl;
        }
      else if(a.isCopy())
        {
          cout << "COPY from " << a.source << endl;
        }
      else assert(0);

      std::set<std::string>::iterator it2;
      for(it2 = a.destlist.begin(); it2 != a.destlist.end(); it2++)
        cout << "  => " << *it2 << endl;
    }
}

RuleSet rules;
Cache::CacheIndex cache;

void cb(const Hash &h, const std::string &url)
{
  cout << "Broken URL " << url << endl;
}

int main()
{
  InstallFinder inf(rules, cache);
  rules.setURLCallback(&cb);

  cout << "Hashes:\n";
  cout << hello << " = HELLO\n";
  cout << dolly << " = DOLLY\n";

  inf.addDep("_file1_hello", hello);
  inf.addDep("_file2_dolly", dolly);
  print(inf);

  rules.addURL(hello, "hello-url");
  rules.reportBrokenURL(hello, "arne");
  print(inf);

  cache.addFile("hello.txt");
  rules.addURL(dolly, "dolly-url");
  print(inf);

  rules.reportBrokenURL(hello, "hello-url");
  rules.reportBrokenURL(hello, "dolly-url");
  print(inf);

  rules.reportBrokenURL(dolly, "dolly-url");
  print(inf);

  return 0;
}
