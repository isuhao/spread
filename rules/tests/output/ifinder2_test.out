Hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF = HELLO
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F = DOLLY

Handled 2 input dependencies into 2 hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : UNHANDLED
  => _file1_hello
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F : UNHANDLED
  => _file2_dolly
Broken URL arne

Handled 2 input dependencies into 2 hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : URL LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF hello-url
  => _file1_hello
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F : UNHANDLED
  => _file2_dolly

Handled 2 input dependencies into 2 hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : COPY from hello.txt
  => _file1_hello
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F : URL z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F dolly-url
  => _file2_dolly
Broken URL hello-url
Broken URL dolly-url

Handled 2 input dependencies into 2 hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : COPY from hello.txt
  => _file1_hello
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F : URL z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F dolly-url
  => _file2_dolly
Broken URL dolly-url

Handled 2 input dependencies into 2 hashes:
LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF : COPY from hello.txt
  => _file1_hello
z5uehbmaSdLx6w2YGCXGTwadrwpgxfwZ-HRb9ePSkJ4F : UNHANDLED
  => _file2_dolly
